desc/objectives/linear_objectives.py

```diff
@@ -142,66 +137,34 @@ def build(self, use_jit=False, verbose=1):
-                f"couldn't find parameter {par} in optimizable_parameters: "
-                + f"{thing.optimizable_params}",
-            )
-        self._params = params
 
-        # replace indices=True with actual indices
-        if isinstance(self._indices, bool) and self._indices:
-            self._indices = [np.arange(thing.dimensions[par]) for par in self._params]
-        # make sure its iterable if only a scalar was passed in
-        if not isinstance(self._indices, (list, tuple)):
-            self._indices = [self._indices]
-        # replace idx=True with array of all indices, throwing an error if the length
-        # of indices is different from number of params
-        indices = {}
-        errorif(
-            len(self._params) != len(self._indices),
-            ValueError,
-            f"not enough indices ({len(self._indices)}) "
-            + f"for params ({len(self._params)})",
-        )
-        for idx, par in zip(self._indices, self._params):
-            if isinstance(idx, bool) and idx:
-                idx = np.arange(thing.dimensions[par])
-            indices[par] = np.atleast_1d(idx)
-        self._indices = indices
-        self._dim_f = sum(t.size for t in self._indices.values())
-
-        default_target = {
-            par: thing.params_dict[par][self._indices[par]] for par in params
-        }
-        default_bounds = None
-        target, bounds = self._parse_target_from_user(
-            self._target_from_user, default_target, default_bounds, indices
-        )
-        if target:
-            self.target = jnp.concatenate([target[par] for par in params])
-            self.bounds = None
-        else:
-            self.target = None
-            self.bounds = (
-                jnp.concatenate([bounds[0][par] for par in params]),
-                jnp.concatenate([bounds[1][par] for par in params]),
+        # default params
+        default_params = tree_map(lambda dim: np.arange(dim), thing.dimensions)
+        self._params = setdefault(self._params, default_params)
+        self._params = broadcast_tree(self._params, default_params)
+        self._indices = tree_leaves(self._params)
+        assert tree_structure(self._params) == tree_structure(default_params)
+
+        self._dim_f = sum(idx.size for idx in self._indices)
+
+        # default target
+        if self.target is None and self.bounds is None:
+            self.target = np.concatenate(
+                [
+                    np.atleast_1d(param[idx])
+                    for param, idx in zip(tree_leaves(thing.params_dict), self._indices)
+                ]
             )
+
         super().build(use_jit=use_jit, verbose=verbose)
 
     def compute(self, params, constants=None):
         """Compute fixed degree of freedom errors.
 
         Parameters
         ----------
-        params : dict
-            Dictionary of equilibrium degrees of freedom, eg Equilibrium.params_dict
+        params : list of dict
```

Comment:
It has to be able to be broadcast to the same pytree structure as `thing.params_dict`. So the base layer has to be a dictionary of the form `{"param": np.array(indices)}`. Note that the indices have to be NumPy arrays, not lists, because the former is considered a pytree "leaf" and the latter is not. A single dict will be broadcast to the appropriate tree structure, otherwise it has to be a nested list of dicts.